File: mcrt.c
#include<stdio.h>
#include<stdlib.h>
#include<math.h>
#include<time.h>
#define MUBINS 10
#define NPHOTONS 1000000
#define ALBEDO 1.0
#define TAUMAX 10.0

/*
 This program calculates the emergent intensity from a plane parallel slab of optical depth TAUMAX, using NPHOTONS photon packets.
 */
/*==============================================================*/
/*---------------------Struct definitions-----------------------*/
/*==============================================================*/
typedef struct {float cost, phi, sint, cosp, sinp, x, y, z;}photon;
typedef struct {int bins[MUBINS];float theta[MUBINS];}hist;
/*==============================================================*/
/*------------------Function declarations-----------------------*/
/*==============================================================*/
hist *binphotons(float mu[]);
photon *newphoton();
void *scatter(photon *p);
void *intensity(hist *binned, float *p);
/*==============================================================*/
/*------------Main loop, loop over stellar photons--------------*/
/*==============================================================*/
int main(int argc, char *argv[]){
	int i;
	static float mu[NPHOTONS];
	static float intens[MUBINS];
	float *intensp;
	hist workspace;
	//workspace.bins[MUBINS]={0};
	hist *histp;
	static photon test;
	photon *working;
	working = &test;
	histp = &workspace;
	intensp = &intens;
	FILE *file;
	srand(time(NULL));
	/* Initializing works.*/
	for(i=0;i<=NPHOTONS;i++){
		*working = *newphoton();
		*scatter(working);
		mu[i]=working->cost;
		if(i%10000==0){
			printf("%d photons completed\n",i);
		}
	}
	*histp = *binphotons(mu);
	intensity(histp, intensp);
	file=fopen("intensity-c.dat","a+");
	fprintf(file,"%s","#intensity, theta\n");
	for(i=0;i<10;i++){
		fprintf(file,"%f\t%f\n",intens[i],histp->theta[i]);
	}
	fclose(file); /*done!*/ 
	return 0;
}
/*==============================================================*/
/*------------------Function definitions------------------------*/
/*==============================================================*/

/*-----------------------New photons----------------------------*/
photon *newphoton(){
	photon *new;
	photon test;
	new = &test;
	float tmp;
	new->cost = sqrt((float)rand()/RAND_MAX); 				//cos theta
	new->phi = 2.0*M_PI*(float)rand()/RAND_MAX;			// phi
	tmp = new->cost;
	new->sint =  sqrt(1-tmp*tmp); 				//sin theta
	new->cosp = cos(new->phi);						//cos PHI
	new->sinp = sin(new->phi);						//sin phi
	new->x = 0.0;							// x
	new->y = 0.0;							// y
	new->z = 0.0;							// z
	return new;
}

/*-------------------Scatter photons----------------------------*/
void *scatter(photon *scatterptr){
	float tau,s;
	while (scatterptr->z>=0.0 && scatterptr->z<=1.0){
		tau = -1.0*log((float)rand()/RAND_MAX);
		s=tau/TAUMAX;
		scatterptr->x +=s*scatterptr->sint*scatterptr->cosp;			//x
		scatterptr->y +=s*scatterptr->sint*scatterptr->sinp;			//y
		scatterptr->z +=s*scatterptr->cost;					//z
		if (scatterptr->z<0.0){
		*scatterptr = *newphoton();
		}
		if ((float)rand()/RAND_MAX < ALBEDO && scatterptr->z < 1.0){
			scatterptr->cost = 2.0*((float)rand()/RAND_MAX) - 1.0;			//cos theta
			scatterptr->phi = 2.0*M_PI*((float)rand()/RAND_MAX);			//phi
			scatterptr->sint = sqrt(1-(scatterptr->cost)*(scatterptr->cost));			//sin theta
			scatterptr->cosp = cos(scatterptr->phi);					//cos phi
			scatterptr->sinp = sin(scatterptr->phi);					//sin phi
		}
	}	
	return 0;
}

/*-----------------------Bin photons----------------------------*/
hist *binphotons(float mu[]){
	float dtheta,width;
	hist binspace;
	hist *binned;
	binned = &binspace;
	int i,j;
	dtheta = 1.0/MUBINS;
	width = 0.5*dtheta;
	for(i=0;i<MUBINS;i++){
	binned->bins[i]=0;
	binned->theta[i]=acos(i*dtheta+width)*(180.0/M_PI);
	}
	for(i=0;i<NPHOTONS;i++){
		j=abs((int)(mu[i]*MUBINS));
		binned->bins[j]+=1;
	}
	return binned;
}
/*---------------------Calculate intensity-----------------------*/
void *intensity(hist *binned, float *p){
	int i;
	for(i=0;i<=(int)MUBINS;i++){
		p[i] = ((float)binned->bins[i]*(float)MUBINS)/(2.0*(float)NPHOTONS*cos(binned->theta[i]*M_PI/180.0));
	}
	return 0;
}

File: mcrt.py
import numpy as np
import matplotlib.pylab as plt
import matplotlib.cm as cm
from mpl_toolkits.mplot3d import Axes3D
from time import time
import sys

def newphoton():
	cost = np.sqrt(np.random.random_sample())
	phi = 2.0 * np.pi * np.random.random_sample()
	sint = np.sqrt(1-cost*cost)
	cosp = np.cos(phi)
	sinp = np.sin(phi)
	xt, yt, zt = 0.0,0.0,0.0
	return cost, sint, cosp, sinp, xt, yt, zt
	
def scatter():
	cost = 2.0*np.random.random_sample()-1.0
	sint = np.sqrt(1.0-cost*cost)
	phi = 2.0*np.pi*np.random.random_sample()
	cosp = np.cos(phi)
	sinp = np.sin(phi)
	return cost, sint, cosp, sinp, phi

def binphotons(mu, mubins,n):#bin the photons uniformly from 0-90 degrees.
	dtheta=1.0/mubins
	width=0.5*dtheta #see ken wood's slab code.
	theta=np.zeros(mubins)
	bins=np.zeros(mubins)
	for i in range(mubins):
		theta[i]=np.arccos(i*dtheta+width)*(180.0/np.pi)
	for i in range(int(n)):
		#if mu[i]>=0:
		j=abs(int(mu[i]*mubins))
		bins[j]=bins[j]+1
	return bins, theta

if __name__ == "__main__":
	start = time()
	taumax = 71.0    
	nphotons = float(sys.argv[1])
	mubins = 10
	phibins = 10
	mu1=np.array(np.zeros(nphotons))
	phi1=np.array(np.zeros(nphotons))
	i=1
	albedo = 1
	while i <= nphotons:
		cost, sint, cosp, sinp, xt, yt, zt = newphoton()
		###get scattered on
		while zt>=0.0 and zt<=1.0:
			x=xt
			y=yt
			z=zt
			tau=-1.0*np.log(np.random.random_sample())
			s=tau/taumax
			xt=x+s*sint*cosp
			yt=y+s*sint*sinp
			zt=z+s*cost
			if np.random.random_sample() < albedo and zt>0.0 and zt<1.0:
				cost,sint,cosp,sinp,phi = scatter()
		if zt>=1.0:
			mu1[i-1]=cost
			phi1[i-1]=phi 
			i=i+1
	muhist,theta=binphotons(mu1,mubins,nphotons)
	intensity = np.zeros(mubins)
	for j in range(mubins):
		intensity[j]=(muhist[j]*mubins)/(2*sum(muhist)*np.cos(theta[j]*np.pi/180.0))
	print "%d scatterings completed in %f s" % (sum(muhist), (time()-start))
	fig=plt.figure(figsize=(8,8))
	plt.ylabel(r'$\frac{I_\nu}{B_\nu}$')
	plt.xlabel(r'$\theta$')
	plt.plot(theta,intensity,'bo')
	f=open('intensity.dat','w')
	for i in range(theta.size):
		f.write("%4.2f\t%6.4f\n" % (theta[i], intensity[i]))
	f.close()
	fig.savefig('intensity.png')
	fig.clf()
	plt.plot(theta,muhist/nphotons,'bo')
	fig.savefig('energy.png')
